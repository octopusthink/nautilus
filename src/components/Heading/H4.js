import { css } from '@emotion/core';
import styled from '@emotion/styled';
import React from 'react';

import Heading from './';

// We define `<H4>` for developer convenience.
export const Heading4 = (props) => {
  return <Heading level={4} {...props} />;
};

Heading4.defaultProps = {
  children: undefined,
};

Heading4.propTypes = { ...Heading.PropTypes, level: undefined };

export default styled(Heading4)``;
