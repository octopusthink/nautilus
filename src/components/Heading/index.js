import { css } from '@emotion/core';
import invariant from 'invariant';
import PropTypes from 'prop-types';
import React from 'react';

import { useTheme } from 'themes';
import { headingLarge, headingMedium, headingSmall } from 'themes/mixins';

export const Heading = ({
  children,
  level,
  large,
  medium,
  small,
  ...otherProps
}) => {
  invariant(level, 'Heading component requires a level prop');

  const HeadingElement = `h${level}`;
  const theme = useTheme();

  return (
    <HeadingElement
      css={css`
        ${css(headingMedium(theme))};
        margin: 0 0 ${theme.spacing.margin.m};

        ${large &&
          css`
            ${css(headingLarge(theme))}
          `}

        ${small &&
          css`
            ${css(headingSmall(theme))}
          `}
      `}
      {...otherProps}
    >
      {children}
    </HeadingElement>
  );
};

export const HeadingLevels = [1, 2, 3, 4, 5, 6];

Heading.defaultProps = {
  children: undefined,
  level: 2,
  large: true,
  medium: false,
  small: false,
};

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Semantic hierarchy level of the `<h>` element in the markup (ex: `<h3>`) */
  level: PropTypes.oneOf(HeadingLevels).isRequired,
  /** Visual prominency of the heading. Small, medium, large. */
  large: PropTypes.bool,
  medium: PropTypes.bool,
  small: PropTypes.boolm,
};

export default Heading;
