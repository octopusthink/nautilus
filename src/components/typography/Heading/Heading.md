## Available Sizes
### XXLarge Heading
```js
<Heading level={2} size="xxlarge">This is a heading</Heading>
```

### XLarge Heading
```js
<Heading level={2} size="xlarge">This is a heading</Heading>
```

### Large Heading
```js
<Heading level={2} size="large">This is a heading</Heading>
```

### Medium Heading
```js
<Heading level={2}>This is a heading</Heading>
```

### Small Heading
```js
<Heading level={2} size="small">This is a heading</Heading>
```

### XSmall Heading
```js
<Heading level={2} size="xsmall">This is a heading</Heading>
```

### XXSmall Heading
```js
<Heading level={2} size="xxsmall">This is a heading</Heading>
```

## Colors
### Headings with custom colors
```js
<Heading level={2} color="light">This is a heading</Heading>
<Heading level={2} color="medium">This is a heading</Heading>
<Heading level={2} color="dark">This is a heading</Heading>
<Heading level={2} color="accent">This is a heading</Heading>
<Heading level={2} color="danger">This is a heading</Heading>
<Heading level={2} color="success">This is a heading</Heading>
```
