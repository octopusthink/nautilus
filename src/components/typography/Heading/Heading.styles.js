import styled from 'styled-components';

import typography from '../../../base/typography/typography';
import { semanticColors, typographicColors } from '../../../base/color/colorVariables';

export const headingSizes = {
  xxlarge: typography.headingXXLarge,
  xlarge: typography.headingXLarge,
  large: typography.headingLarge,
  medium: typography.headingMedium,
  small: typography.headingSmall,
  xsmall: typography.headingXSmall,
  xxsmall: typography.headingXXSmall,
};

export const headingColors = {
  light: typographicColors.typeLight,
  medium: typographicColors.typeMedium,
  dark: typographicColors.typeDark,
  accent: typographicColors.typeAccent,
  danger: semanticColors.colorDanger,
  success: semanticColors.colorSuccess,
};

const HeadingBase = styled.span`
  margin: 0;
  color: ${p => headingColors[p.color]};
  ${p => headingSizes[p.size]};
`;

export const headingElements = {
  h1: HeadingBase.withComponent('h1'),
  h2: HeadingBase.withComponent('h2'),
  h3: HeadingBase.withComponent('h3'),
  h4: HeadingBase.withComponent('h4'),
  h5: HeadingBase.withComponent('h5'),
  h6: HeadingBase.withComponent('h6'),
};
