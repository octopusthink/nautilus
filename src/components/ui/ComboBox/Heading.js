import { css } from '@emotion/core';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { toUnits } from '../../../styles';
import { useTheme } from '../../../themes';
import ListItem from '../List/Item';

export const ComponentClassName = 'Nautilus-ComboBoxHeading';

export const Heading = (props) => {
  const { children, className, unstyled, ...otherProps } = props;

  const theme = useTheme();

  return (
    <ListItem
      className={classnames(ComponentClassName, className)}
      css={
        unstyled
          ? undefined
          : css`
              background: ${theme.colors.text.default};
              color: ${theme.colors.text.inverse};
              line-height: 2.4rem;
              padding: ${toUnits(theme.spacing.padding.medium)};
              width: 100%;
            `
      }
      unstyled
      {...otherProps}
    >
      {children}
    </ListItem>
  );
};

Heading.defaultProps = {
  children: undefined,
  className: undefined,
  unstyled: false,
};

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  className: PropTypes.string,
  /* @ignore Don't output any CSS styles. */
  unstyled: PropTypes.bool,
};

export const { defaultProps, propTypes } = Heading;

export default Heading;
