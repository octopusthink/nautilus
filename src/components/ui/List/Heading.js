import React from 'react';

export const ListHeading = (props) => {
  return <React.Fragment {...props} />;
};

export default ListHeading;
