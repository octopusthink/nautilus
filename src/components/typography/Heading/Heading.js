import PropTypes from 'prop-types';
import React from 'react';

import { headingSizes, headingColors, headingElements } from './Heading.styles';

export const Heading = ({ children, color, level, size, ...others }) => {
  const HeadingElement = headingElements[`h${level}`];
  return (
    <HeadingElement
      color={color}
      size={size}
      {...others}
    >
      {children}
    </HeadingElement>
  );
};

export const HeadingColors = Object.keys(headingColors);
export const HeadingLevels = [1, 2, 3, 4, 5, 6];
export const HeadingSizes = Object.keys(headingSizes);

Heading.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Color of the Heading. Options are 'dark', 'medium', 'light', 'accent', 'danger', and 'success' */
  color: PropTypes.oneOf(HeadingColors),
  /** Hierarchy level of the Heading (ex: h3) */
  level: PropTypes.oneOf(HeadingLevels),
  /** Size of the Heading. Options are 'xxsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge' */
  size: PropTypes.oneOf(HeadingSizes),
};

Heading.defaultProps = {
  color: 'dark',
  level: 2,
  size: 'medium',
};
