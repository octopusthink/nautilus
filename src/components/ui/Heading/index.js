import { css } from '@emotion/core';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { ComponentClassName as ListClassName } from 'components/ui/List';
import { ComponentClassName as ParagraphClassName } from 'components/ui/Paragraph';
import { heading, toUnits } from 'styles';
import { useTheme } from 'themes';

const LARGE = 2;
const MEDIUM = 3;
const SMALL = 4;

const ComponentClassName = 'Nautilus-Heading';
const HeadingLevels = [LARGE, MEDIUM, SMALL];

export const headingClassNameForLevel = (level) => {
  return `${ComponentClassName}-${level}`;
};

export const Heading = (props) => {
  const theme = useTheme();
  const { children, className, level, ...otherProps } = props;
  const HeadingElement = `h${level}`;

  return (
    <HeadingElement
      className={classnames(headingClassNameForLevel(level), className)}
      css={css`
        margin: 0 0 ${toUnits(theme.spacing.margin.medium)};
        .${ListClassName} + &,
        .${ParagraphClassName} + &,
        /* TODO: Replace these with actual Nautilus components */
        dl + & {
          margin-top: ${toUnits(theme.spacing.margin.medium)};
        }

        ${level === SMALL && heading.small(theme)};
        ${level === MEDIUM && heading.medium(theme)};
        ${level === LARGE && heading.large(theme)};
      `}
      {...otherProps}
    >
      {children}
    </HeadingElement>
  );
};

Heading.defaultProps = {
  className: undefined,
  children: undefined,
  level: 2,
};

Heading.propTypes = {
  /** @ignore */
  className: PropTypes.string,
  /** @ignore */
  children: PropTypes.node,
  /** Semantic hierarchy level of the `<h>` element in the markup (ex: `<h3>`). The more semantically important the level, the larger the heading will appear visually; an `<h2>` will be visually styled as "large" while an `<h4>` will be visually small. */
  level: PropTypes.oneOf(HeadingLevels),
};

export default Heading;
