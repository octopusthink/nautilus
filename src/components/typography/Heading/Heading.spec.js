import React from 'react';

import { Heading, HeadingSizes, HeadingColors, HeadingLevels } from './Heading';

describe('Heading Component', () => {
  context('when nothing specified', () => {
    it('should render correctly', () => {
      const subject = shallow(<Heading>Hello</Heading>);
      expect(subject).toMatchSnapshot();
    });
  });

  HeadingSizes.forEach((size) => {
    context(`when @size specified as ${size}`, () => {
      it(`renders a Heading size of ${size}`, () => {
        const subject = mount(<Heading size={size}>Hello</Heading>);
        expect(subject.prop('size')).toBe(size);
      });
    });
  });

  HeadingLevels.forEach((level) => {
    context(`when @level specified as ${level}`, () => {
      it(`renders a Heading level of ${level}`, () => {
        const subject = mount(<Heading level={level}>Hello</Heading>);
        expect(subject.prop('level')).toBe(level);
      });
    });
  });

  HeadingColors.forEach((color) => {
    context(`when @color specified as ${color}`, () => {
      it(`renders a Heading color of ${color}`, () => {
        const subject = mount(<Heading color={color}>Hello</Heading>);
        expect(subject.prop('color')).toBe(color);
      });
    });
  });
});
