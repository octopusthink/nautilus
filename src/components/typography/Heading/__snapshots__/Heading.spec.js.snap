// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading Component when nothing specified should render correctly 1`] = `
<ForwardRef
  color="dark"
  size="medium"
>
  Hello
</ForwardRef>
`;
